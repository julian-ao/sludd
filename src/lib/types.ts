export type LocationQueryData = {
    metadata?: {
        totaltAntallTreff: number;
    };
    navn: LocationData[];
};

export type LocationData = {
    representasjonspunkt: {
        nord: number;
        øst: number;
    };
    stedsnavn: {
        skrivemåte: string;
    }[];
    stedsnummer: string;
    navneobjekttype?: string;
    kommuner?: {
        kommunenummer?: string;
        kommunenavn?: string;
    }[];
};

export type WeatherQueryData = {
    properties: {
        meta: {
            updated_at: string;
        };
        timeseries: {
            time: string;
            data: {
                instant: {
                    details: {
                        air_temperature: number;
                        wind_speed: number;
                        relative_humidity: number;
                        wind_from_direction: number;
                    };
                };
                next_1_hours: {
                    summary: {
                        symbol_code: string;
                    };
                    details: {
                        precipitation_amount: number;
                    };
                };
            };
        }[];
    };
};
