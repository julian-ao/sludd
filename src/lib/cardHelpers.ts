// Dictionary of weather descriptions for the different weather types
const weatherDescription: { [key: string]: string } = {
    clearsky: "Clear sky",
    fair: "Fair",
    partlycloudy: "Partly cloudy",
    cloudy: "Cloudy",
    lightrainshowers: "Light rain showers",
    rainshowers: "Rain showers",
    heavyrainshowers: "Heavy rain showers",
    lightrainshowersandthunder: "Light rain showers and thunder",
    rainshowersandthunder: "Rain showers and thunder",
    heavyrainshowersandthunder: "Heavy rain showers and thunder",
    lightsleetshowers: "Light sleet showers",
    sleetshowers: "Sleet showers",
    heavysleetshowers: "Heavy sleet showers",
    lightssleetshowersandthunder: "Light sleet showers and thunder",
    sleetshowersandthunder: "Sleet showers and thunder",
    heavysleetshowersandthunder: "Heavy sleet showers and thunder",
    lightsnowshowers: "Light snow showers",
    snowshowers: "Snow showers",
    heavysnowshowers: "Heavy snow showers",
    lightssnowshowersandthunder: "Light snow showers and thunder",
    snowshowersandthunder: "Snow showers and thunder",
    heavysnowshowersandthunder: "Heavy snow showers and thunder",
    lightrain: "Light rain",
    rain: "Rain",
    heavyrain: "Heavy rain",
    lightrainandthunder: "Light rain and thunder",
    rainandthunder: "Rain and thunder",
    heavyrainandthunder: "Heavy rain and thunder",
    lightsleet: "Light sleet",
    sleet: "Sleet",
    heavysleet: "Heavy sleet",
    lightsleetandthunder: "Light sleet and thunder",
    sleetandthunder: "Sleet and thunder",
    heavysleetandthunder: "Heavy sleet and thunder",
    lightsnow: "Light snow",
    snow: "Snow",
    heavysnow: "Heavy snow",
    lightsnowandthunder: "Light snow and thunder",
    snowandthunder: "Snow and thunder",
    heavysnowandthunder: "Heavy snow and thunder",
    fog: "Fog",
};

// Takes a symbol_code and returns the corresponding weather description
export function parseWeatherDescription(symbol_code: string) {
    const splitSymbolCode = symbol_code.split("_");
    const description = weatherDescription[splitSymbolCode[0]];

    return description;
}
// Dictionary of colors for the different weather types
const weatherColors = {
    clearsky: ["#ded364", "#ede26f", "#d9cf64"],
    fair: ["#8FEFBE", "#90F0AE", "#90F0CF"],
    cloudy: ["#8C9C9A", "#8D9C8E", "#9C998C"],
    rain: ["#8AD2EA", "#A4DCEE", "#93d1e6"],
    snow: ["#6E8FF5", "#6EA7F5", "#6EBEF5"],
    sleet: ["#9FA1ED", "#AB9FEC", "#B89FED"],
    fog: ["#5E6385", "#605E84", "#675E85"],
};

// Takes a symbol_code and returns the corresponding color
export function getColorFromWeatherDescription(symbol_code: string) {
    //use weatherColors to get color from symbol_code
    for (const [key, value] of Object.entries(weatherColors)) {
        if (symbol_code.includes(key)) {
            if (Array.isArray(value)) {
                return value[Math.floor(Math.random() * value.length)];
            }
            return value;
        }
    }
    return "#000000";
}
