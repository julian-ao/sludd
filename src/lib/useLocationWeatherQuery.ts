import { useEffect, useState } from "react";
import { useQuery } from "@tanstack/react-query";
import { LocationQueryData, WeatherQueryData } from "./types";

interface locationName {
    locationName: string;
    filter?: string;
}

interface locationId {
    locationId: string;
    filter?: string;
}

// Custom hook for fetching weather data based on locationName or locationId
export const useLocationWeatherQuery = (data: locationName | locationId) => {

    const [locationData, setLocationData] = useState<LocationQueryData | null>(
        null,
    );

    // Fetch location data (cordinates) based on locationName or locationId
    const locationQuery = useQuery<LocationQueryData, unknown>(
        ["location", data],
        async () => {
            if ("locationId" in data && data.locationId) {
                const url = `https://ws.geonorge.no/stedsnavn/v1/sted?stedsnummer=${
                    data.locationId
                }&utkoordsys=4258${
                    data.filter ? `&filtrer=${data.filter}` : ""
                }`;
                const res = await fetch(url);
                return res.json();
            } else if ("locationName" in data) {
                const base_url = `https://ws.geonorge.no/stedsnavn/v1/sted?sok=${data.locationName}&utkoordsys=4258&treffPerSide=1&side=1&filtrer=navn.representasjonspunkt,navn.stedsnavn.skrivemåte,navn.navneobjekttype,navn.stedsnummer`;
                const res = await fetch(base_url);
                return res.json();
            }

            throw new Error("Invalid data");
        },
    );

    // Set locationData when locationQuery is successful
    useEffect(() => {
        if (locationQuery.isSuccess) {
            setLocationData(locationQuery.data!);
        }
    }, [locationQuery]);

    // Fetch weather data based on cords from locationData
    const weatherQuery = useQuery<WeatherQueryData, unknown>(
        ["weather", locationData?.navn[0].representasjonspunkt, location],
        async () => {
            const res = await fetch(
                `https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=${locationData?.navn[0].representasjonspunkt.nord}&lon=${locationData?.navn[0].representasjonspunkt.øst}`,
            );
            return res.json();
        },
        {
            enabled: !!locationData,
        },
    );

    return { locationData, weatherQuery };
};
