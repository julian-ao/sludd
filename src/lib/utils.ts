// Convert date to readable format
export const convertDateToReadable = (date: string) => {
    const month = date.slice(5, 7);
    const day = parseInt(date.slice(8, 10), 10).toString(); // Remove leading zero
    const hour = date.slice(11, 13);
    const minute = date.slice(14, 16);

    return `${day}. ${months[
        Number(month) - 1
    ].toLocaleLowerCase()} ${hour}:${minute}`;
};

const months = [
    "Januar",
    "Februar",
    "Mars",
    "April",
    "Mai",
    "Juni",
    "Juli",
    "August",
    "September",
    "Oktober",
    "November",
    "Desember",
];
