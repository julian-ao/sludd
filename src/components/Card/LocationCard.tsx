import './Card.css';
import { useLocationWeatherQuery } from '../../lib/useLocationWeatherQuery';
import { getColorFromWeatherDescription, parseWeatherDescription } from '../../lib/cardHelpers';
import CardSkeleton from './CardSkeleton';


const LocationCard = ({ location }: { location: string }) => {
    const { locationData, weatherQuery } = useLocationWeatherQuery(location);

    if (weatherQuery.isLoading || !locationData) {
        return <CardSkeleton />;
    }

    if (weatherQuery.isError) {
        return <CardSkeleton weatherDescription="Error" locationName='Not found' />;
    }

    const data = weatherQuery.data.properties.timeseries[0].data;
    const symbol_code = data.next_1_hours.summary.symbol_code;
    const temperature = data.instant.details.air_temperature;

    const weatherDescription = parseWeatherDescription(symbol_code);
    const weatherColor = getColorFromWeatherDescription(symbol_code);

    return (
        <CardSkeleton
            weatherColor={weatherColor}
            temperature={temperature}
            locationName={locationData?.navn[0].stedsnavn[0].skrivemÃ¥te}
            symbol_code={symbol_code}
            weatherDescription={weatherDescription}
        />
    );
};

export default LocationCard;
