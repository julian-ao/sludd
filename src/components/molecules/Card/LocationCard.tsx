import './Card.css';
import { getColorFromWeatherDescription } from '../../../lib/cardHelpers';
import CardSkeleton from './CardSkeleton';
import { LocationData, WeatherQueryData } from '../../../lib/types';
import { FC } from 'react';
import { useQuery } from '@tanstack/react-query';

interface LocationCardProps {
    locationData: LocationData;
}

const LocationCard: FC<LocationCardProps> = ({ locationData }) => {
    const coordinates = locationData.representasjonspunkt;
    // const weatherQuery = useWeather(coordinates.nord, coordinates.øst);
    const url = `https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=${coordinates.nord}&lon=${coordinates.øst}`;
    const { data: weatherData, isLoading: weatherIsLoading, isError: weatherIsError } = useQuery<WeatherQueryData>(
        ["weather", coordinates.nord, coordinates.øst],
        async () => {
            const res = await fetch(url);
            if (!res.ok) {
                console.log("res:", res);
                throw new Error("Network response was not ok");
            }
            return res.json();
        },
        {
            enabled: !!coordinates
        },
    );


    if (weatherIsLoading || !locationData) {
        return <CardSkeleton />;
    }

    if (weatherIsError) {
        return <CardSkeleton locationName='Not found' />;
    }

    const weatherProperties = weatherData.properties.timeseries[0].data;
    const symbol_code = weatherProperties.next_1_hours.summary.symbol_code;
    const temperature = weatherProperties.instant.details.air_temperature;

    const weatherColor = getColorFromWeatherDescription(symbol_code);

    return (
        <CardSkeleton
            weatherColor={weatherColor}
            temperature={temperature}
            locationName={locationData?.stedsnavn[0].skrivemåte}
            locationId={parseInt(locationData?.stedsnummer)}
            symbol_code={symbol_code}
            navneobjekttype={locationData?.navneobjekttype}
        />
    );
};

export default LocationCard;
