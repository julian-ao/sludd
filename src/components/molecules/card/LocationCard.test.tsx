import { cleanup, render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { vi } from 'vitest';
import { WeatherQueryData } from '../../../lib/types';
import LocationCard from './LocationCard';

const wrapper = ({ children }: React.PropsWithChildren<object>) => (
    <BrowserRouter>
        {children}
    </BrowserRouter>
)

describe('LocationCard', () => {
    let queryClient: QueryClient;

    beforeEach(() => {
        queryClient = new QueryClient();
    });

    afterEach(() => {
        vi.restoreAllMocks();
        cleanup();
    });

    it('render LocationCard with mock response', async () => {
        const mockResponse = {
            properties: {
                meta: {
                    updated_at: '2021-09-27T12:00:00Z',
                },
                timeseries: [
                    {
                        time: '2021-09-27T12:00:00Z',
                        data: {
                            instant: {
                                details: {
                                    air_temperature: 10,
                                    wind_speed: 5,
                                    relative_humidity: 10,
                                    wind_from_direction: 10,
                                },
                            },
                            next_1_hours: {
                                summary: {
                                    symbol_code: 'clearsky_day',
                                },
                                details: {
                                    precipitation_amount: 0,
                                },
                            },
                        },
                    },
                ],
            },
        } as WeatherQueryData;

        vi.spyOn(global, 'fetch').mockResolvedValueOnce({
            json: vi.fn().mockResolvedValueOnce(mockResponse),
        } as never);

        vi.spyOn(Math, 'random').mockReturnValue(0.4);

        const { container } = render(
            <QueryClientProvider client={queryClient}>
                <LocationCard locationData={{
                    representasjonspunkt: {
                        nord: 0,
                        øst: 0
                    },
                    stedsnavn: [{ skrivemåte: 'test' }],
                    stedsnummer: 'test',
                }} />
            </QueryClientProvider>,
            { wrapper },
        );
        expect(container).toMatchSnapshot();
    });
});