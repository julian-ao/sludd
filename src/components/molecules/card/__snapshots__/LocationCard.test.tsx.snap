// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LocationCard > render LocationCard with mock response 1`] = `
<div>
  <div
    class="card"
    style="background-color: white;"
  >
    <div
      class="card-content"
    >
      <div
        class="header-section"
      >
        <p
          class="location-text"
        >
          Laster...
        </p>
        <p
          class="type-text"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-location-dot "
            data-icon="location-dot"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 5px;"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
              fill="currentColor"
            />
          </svg>
          , 
        </p>
      </div>
      <div
        class="wave-section"
        style="background-color: white;"
      >
        <div
          style="background: url(data:image/svg+xml;utf8,%3Csvg%20viewBox%3D%220%200%201000%20126%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Cpath%20d%3D%22M0%2042.9533C178.148%20-12.5894%20287.4%20-13.7474%20500%2042.9533C682.727%2088.1203%20798.763%2097.7368%201000%2042.9533V125.861H0V42.9533Z%22%20fill%3D%22white%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E); position: relative; bottom: 0px; width: 100%; height: 30px; background-size: 300px 40px; animation: wave 21s linear infinite reverse; opacity: 1;"
        />
        <div
          class="description-section positive-temperature"
        >
          <div
            class="loading"
          />
          <div
            style="display: flex; align-items: center;"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-temperature-three-quarters "
              data-icon="temperature-three-quarters"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="margin-right: 5px;"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M160 64c-26.5 0-48 21.5-48 48V276.5c0 17.3-7.1 31.9-15.3 42.5C86.2 332.6 80 349.5 80 368c0 44.2 35.8 80 80 80s80-35.8 80-80c0-18.5-6.2-35.4-16.7-48.9c-8.2-10.6-15.3-25.2-15.3-42.5V112c0-26.5-21.5-48-48-48zM48 112C48 50.2 98.1 0 160 0s112 50.1 112 112V276.5c0 .1 .1 .3 .2 .6c.2 .6 .8 1.6 1.7 2.8c18.9 24.4 30.1 55 30.1 88.1c0 79.5-64.5 144-144 144S16 447.5 16 368c0-33.2 11.2-63.8 30.1-88.1c.9-1.2 1.5-2.2 1.7-2.8c.1-.3 .2-.5 .2-.6V112zM208 368c0 26.5-21.5 48-48 48s-48-21.5-48-48c0-20.9 13.4-38.7 32-45.3V144c0-8.8 7.2-16 16-16s16 7.2 16 16V322.7c18.6 6.6 32 24.4 32 45.3z"
                fill="currentColor"
              />
            </svg>
            <p
              style="color: rgb(136, 136, 136);"
            >
              Laster...
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
